### ISSUE TITLE
Memory error on worker can cause dropped keys

### ISSUE THREAD

**Issue by jcrist** [dask/distributed#1088]
Memory errors on worker can cause a key to be dropped from `Worker.data`, resulting in further computation to use the literal key instead of the referenced data. See https://github.com/dask/dask/issues/2361.

**Comment by mrocklin** [comment]
This is probably on me to fix.  If I recall correctly I think that this issue occurs when dumping a piece of data to disk fails.  I think that we silently fail in this case.  Instead we should probably propagate out some informative exception to all dependent tasks.

**Comment by mrocklin** [comment]
(although others are certainly welcome to play with this)

**Comment by mrocklin** [comment]
This reproduces a similar issue:

```python
class Bad(object):
    def __getstate__(self):
        raise Exception()

    def __sizeof__(self):
        return int(100e9)

from dask.distributed import Client
Client().submit(future, Bad).result()
```

**Comment by jakirkham** [comment]
> If I recall correctly I think that this issue occurs when dumping a piece of data to disk fails.

Yeah, in the xref'd issue above I had been writing out results to disk.

**Comment by mrocklin** [comment]
This is being resolved in #1096
